02-intermediate-ruby/references-and-values.md

```diff
@@ -0,0 +1,267 @@
+# References and Values
+
+## Learning Goals
+At the end of this lesson, students should be able to...
+
+- Discuss how objects are stored in Ruby
+- Differentiate between _references_ and _values_
+- Compare _modifying an object_ with _reassigning a variable_
+
+## Motivation
+
+We're going to start today with some Ruby code that does something a little unexpected. It's a method that takes an array of strings as an argument, and truncates (chops off the end of) all the strings with more than three characters. Or at least that's the idea.
+
+```ruby
+def short_strings_1(input)
+  result = []
+  input.each do |word|
+    # Slice characters 0 to 2
+    result << word[0..2]
+  end
+  input = result
+end
+
+pets = ['dog', 'parrot', 'cat', 'llama']
+short_strings_1(pets)
+puts "#{pets}"
+
+```
+
+Running this code results in `["dog", "parrot", "cat", "llama"]`. The array in unchanged! Let's do some debugging:
+
+```ruby
+def short_strings_1(input)
+  result = []
+  input.each do |word|
+    # Slice characters 0 to 2
+    result << word[0..2]
+  end
+  input = result
+  puts "Inside short_strings_1, input is"
+  puts "#{input}"
+end
+
+pets = ['dog', 'parrot', 'cat', 'llama']
+short_strings_1(pets)
+puts "After calling short_strings_1"
+puts "#{pets}"
+```
+
+The results:
+
+```
+Inside short_strings_1, input is
+["dog", "par", "cat", "lla"]
+After calling short_strings_1
+["dog", "parrot", "cat", "llama"]
+```
+
+Seems like our method is indeed creating a list of shortened words, but our outer variable isn't being updated. The reason why has to do with references and values, and how data is stored in a computer.
+
+As an aside: one way to fix our method is to simply return the new array, and when calling it say `pets = short_strings_1(pets)`. However, sometimes this isn't an option - for example, what if our method was supposed to return the number of words that had to be truncated?
+
+## References and Values
+
+When we create an array in Ruby (or a string or a hash or any other complex data type), we're actually creating two things.
+
+The first is the _value_ of the array, which involves asking the operating system for a bit of memory and then putting our data in it. You can think of this as the actual object. Each piece of memory we get from the OS has an _address_ representing its physical location in hardware, which is how we get back to it later.
+
+The second is a _reference_ to the array, which ties together the address of that memory with a name for our program to use. References are sometimes called _pointers_ (especially in C), and we say that a variable _points to_ or _references_ an object.
+
+![references and variables](images/references_and_values/references_and_values.png)
```

Comment:
It's not too clear to me where the 1234 comes into play here (i mean, i know what it is but it doesn't seem clear in the diagram) - so it might be good to put obj ID in parens or something along those lines.  It might also be good to copy the pets array before the diagram so folks can recall what that variable was referring to